### DISCUSSION TITLE
提供等宽字体的自定义选项

### DISCUSSION THREAD

**Discussion by yuuuxt** [F9y4ng/GreasyFork-Scripts#101] [问答专区(Question / Answer)] [answered]
在#74 中提供了预置的等宽字体, 并`从左至右，依次查询你系统上是否安装过以上英文等宽字体，先查到哪个字体就用哪个进行渲染`;

如已安装其中的多个字体, 但想使用的是非是排在首位的字体(如: 同时安装了`Fira Code`和`Source Code Pro`, 想使用`Source Code Pro`), 就不方便了;

建议提供自定义选项, 提供默认值并可修改.

(我现在是直接代码里移除了我不想用但已安装的等宽字体来达到效果)

**Comment by F9y4ng** [comment]
脚本对等宽字体的定义是仅对于非特殊需求的用户所提供的**简单的**等宽字体替换功能。

如果需要更高级的字体设定，建议你将 **排除渲染的HTML标签** 中 `pre,pre *,code,code *` 删除后保存数据，并通过浏览器字体设置的等宽字体设定来定义你需要的个性化字体。

**Reply by yuuuxt** [reply]
> 脚本对等宽字体的定义是仅对于非特殊需求的用户所提供的**简单的**等宽字体替换功能。
> 
> 如果需要更高级的字体设定，建议你将 **排除渲染的 HTML 标签** 中 `pre,pre *,code,code *` 删除后保存数据，并通过浏览器字体设置的等宽字体设定来定义你需要的个性化字体。

似乎不是这样, 如下:
- 使用firefox 浏览器, 在fonts - advanced 中设置等宽字体, 勾选下面的`Allow pages to choose their own fonts, instead of your selections above`
- 按上述移除`pre, ...`
- 结果表现为渲染成非等宽字体 (预期的结果, 因为脚本生效了?)

**Reply by F9y4ng** [reply]
在 **排除渲染的HTML标签** 删除 `pre,pre *,code,code *`后，脚本不会执行插入等宽字体相关的样式。

如果你认为是脚本造成的问题，你可以将脚本禁用，再通过浏览器设定等宽字体看看是否生效，来排除脚本问题。

**Reply by yuuuxt** [reply]
> 在 **排除渲染的 HTML 标签** 删除 `pre,pre *,code,code *` 后，脚本不会执行插入等宽字体相关的样式。
> 
> 如果你认为是脚本造成的问题，你可以将脚本禁用，再通过浏览器设定等宽字体看看是否生效，来排除脚本问题。

刚刚一起测试了, 是脚本生效了, 插入了非等宽字体的样式, 变成了非等宽字体 (补充一下: 字体重写是打开的, 设置了中英字体).

我还是录个屏证明一下吧:
![temp](https://user-images.githubusercontent.com/3153534/177009108-c9de6743-baf8-4d8e-befc-8e7e83f71ce7.gif)

**Reply by yuuuxt** [reply]
以及, 这个问题(选择一个等宽字体)还是有很简单的workaround的, 就是直接改代码里提供的等宽字体列表;

#102 这个更麻烦点, 把github的等宽字体设置成了非等宽字体

**Reply by F9y4ng** [reply]
基本和github是同一个问题，都是样式优先级和默认中文字体参数设定的造成的。

同 #102 的建议一样，关闭字体重写功能，通过浏览器设定英文字体及等宽字体。

**Reply by yuuuxt** [reply]
> 基本和github是同一个问题，都是样式优先级和默认中文字体参数设定的造成的。
> 
> 同 #102 的建议一样，关闭字体重写功能，通过浏览器设定英文字体及等宽字体。

这个(最开始的问题)是小问题啦, 如果没有github的问题的话, 这个是可以忽略的, 代码里改一改就可以了;

上面演示主要还是说`排除渲染的HTML标签 中 pre,pre *,code,code * 删除后保存数据` 这个思路是不对的, 反而会引起类似#102 的问题.

**Reply by F9y4ng** [reply]
检查一下你的本地设置吧，是否使用其他软件对字体进行了热替换等等，我本地均无法复现你所提到的问题。

![屏幕截图 2022-07-03 010757](https://user-images.githubusercontent.com/15855068/177009847-e67bf2fc-0a71-45da-ae29-d71e5f3c4411.png)

**Reply by yuuuxt** [reply]
> 检查一下你的本地设置吧，是否使用其他软件对字体进行了热替换等等，我本地均无法复现你所提到的问题。
> 
> ![屏幕截图 2022-07-03 010757](https://user-images.githubusercontent.com/15855068/177009847-e67bf2fc-0a71-45da-ae29-d71e5f3c4411.png)

你这儿其实已经体现问题了: 代码段是理应是等宽字体, 目前是设置的非等宽字体(Bernard MT); 
我上面表达的也是这个意思, 这是`排除渲染的HTML标签 中 pre,pre *,code,code * 删除后保存数据` 引起的问题

**Reply by F9y4ng** [reply]
`pre,pre *,code,code *` 是强制定义等宽字体的触发条件，删除后就没有等宽字体定义了。

以下是等宽字体Monaco的示例，也不存在你说的问题。
![image](https://user-images.githubusercontent.com/15855068/177010596-04c7cbb0-539f-4089-bb7d-eaf74563c35e.png)

**Reply by yuuuxt** [reply]
> `pre,pre *,code,code *` 是强制定义等宽字体的触发条件，删除后就没有等宽字体定义了。
> 
> 以下是等宽字体Monaco的示例，也不存在你说的问题。 ![image](https://user-images.githubusercontent.com/15855068/177010596-04c7cbb0-539f-4089-bb7d-eaf74563c35e.png)

是的, 我也是这么理解的, 所以说`排除渲染的HTML标签 删除 pre,pre *,code,code *` 不能解决 `已安装其中的多个字体, 但想使用的是非是排在首位的字体` 的问题 - 这个问题不需要改变选择器, 需要改变的是字体样式(这个列表), 而目前虽然不像元素选择和排除这样可以直接填, 也是可以直接改代码的.

**Reply by F9y4ng** [reply]
你想要的是自定义等宽字体的功能，我需要考虑的是如何在有限的空间内塞进更多的文本域且不会让人误操作而造成样式无效的问题。😂

如果仅仅是需要自定义等宽字体，你甚至可以尝试其他字体设置扩展或脚本看看。

**Reply by yuuuxt** [reply]
是的, 只是一个小问题, 直接改一下代码也是ok的, 当然给个框便于自定义更好, 类似现在的这两个选择器 - 一开始表达的就是这个意思😂

我的需求还是自定义中+英+等宽:
- 中英思路同 #83, 
- 等宽有预设, 改一下代码可以进一步自定义 (这个 #102);

目前就是中英自定义有两类问题:
- 选择器不全, 类似 #103 报告的YouTube;
- 影响到了等宽字体, 类似 #102 报告的github;

还是挺不好搞的, 没有一开始想象的完整, 还不能替代我目前的用法(firefox强制设置字体为主).

**Reply by F9y4ng** [reply]
其实挺麻烦，因为不是所有的网站都使用`pre,code`来定义代码块，那要处理的问题就很多，不仅仅是字体定义的问题了，元素的选择、兼容过滤、网站之间的冲突、与代码着色的兼容等等……想想都头疼。

这也是为什么当初对等宽字体这块我很简单的做了处理，没有更深入，第一，小众；第二，非常的麻烦。

毕竟这和处理单一网站的脚本不同，我需要兼容所有网站的各种需求，所以一个简单的小功能，要兼容众多网站，问题就来了。

**Comment by F9y4ng** [comment, marked as answer]
1. **自定义等宽字体**：自定义等宽字体的设置功能（字体排序）会在下个周更新中完成。
2. **样式选择器不全**：需要用户根据自己的实际需求自行在对应的文本域进行填写（需要CSS样式语法基础）。
3. #102 此类非`pre,code`的代码块将不做额外处理，因工作量太大，如有特殊需求，可尝试其他扩展或脚本替代。
